### ISSUE TITLE
Add `outlines.generate.fsm()` API entrypoint

### ISSUE THREAD

**Issue by rlouf** [outlines-dev/outlines#670]
As suggested in https://github.com/outlines-dev/outlines/issues/666#issuecomment-1947685486, we should allow users to pass any DFA (with vocabulary = characters) to generate text that is in the vocabulary of that DFA.

**Comment by lapp0** [comment]
We should allow the passing of an `interegular.FSM` because the desired operations requested in the linked issue, and those requested by users in other issues are not possible with a `RegexFSM`.

We should modify `RegexFSM`, perhaps to have `RegexFSM.from_interegular_fsm`

https://github.com/outlines-dev/outlines/blob/main/outlines/fsm/fsm.py#L94-L125

**Comment by rlouf** [comment]
> We should modify `RegexFSM`, to have `RegexFSM.from_interegular_fsm`

This is a hint that we have a naming problem here. And we might as well fix it until more libraries depend on Outlines' internals.

**Comment by lapp0** [comment]
Maybe `TokenIndexFSM`

**Comment by miftahmoha** [comment]
Can someone please explain to me, I'm confused. 

Does the user have to create the FSM with `interegular` and pass it to a method `outlines.generate.fsm(fsm=interegular_fsm)` which returns the `SequenceGenerator`?

I guess `RegexFSM.from_interegular_fsm` should turn an `interegular_fsm` object into a `RegexFSM` object so that it can be processed by `SequenceGenerator`?

I'm interested in working on it if it's possible.

**Comment by lapp0** [comment]
>Does the user have to create the FSM with interegular and pass it to a method outlines.generate.fsm(fsm=interegular_fsm) which returns the SequenceGenerator?

Yes. `interegular` allows you to perform helpful operations on FSMs - concatenation, OR / XOR, intersection, negation, etc. `RegexFSM` doesn't have these capabilities.  

>I guess RegexFSM.from_interegular_fsm should turn an interegular_fsm object into a RegexFSM object so that it can be processed by SequenceGenerator?

Yes exactly!

>I'm interested in working on it if it's possible.

That's great news! 

[The current `RegexFSM.__init__` logic](https://github.com/outlines-dev/outlines/blob/main/outlines/fsm/fsm.py#L94-L125) involves conversion from pattern -> `interegular.FSM` -> initialization of Outlines `RegexFSM`. Ideally `from_interegular_fsm` wouldn't repeat logic with `__init__`, but instead they'd share as much logic as possible using a helper function.

I'd work off of https://github.com/outlines-dev/outlines/blob/main/outlines/generate/regex.py#L10-L36 for the second step, creating `outlines.generate.fsm()`

Please ping me if you have any questions.

**Comment by rlouf** [comment]
That seems reasonable :)

**Comment by miftahmoha** [comment]
@lapp0 Done.

It seems that it is the case, a different run _(same input)_ implies a different hash for `interegular.fsm.FSM` object.

**Comment by lapp0** [comment]
Should be closed thanks to @miftahmoha's implementation and great documentation!

https://github.com/outlines-dev/outlines/pull/699
